pg/src/main/java/org/jdbi/v3/pg/ArrayColumnMapper.java

```diff
@@ -0,0 +1,70 @@
+package org.jdbi.v3.pg;
+
+import java.sql.Array;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.SQLFeatureNotSupportedException;
+import java.util.concurrent.CopyOnWriteArraySet;
+
+import org.jdbi.v3.StatementContext;
+import org.jdbi.v3.tweak.ResultColumnMapper;
+
+public class ArrayColumnMapper implements ResultColumnMapper<Object[]> {
+    private static final CopyOnWriteArraySet<Integer> UNSUPPORTED_TYPES = new CopyOnWriteArraySet<>();
+    private final Class<?> componentType;
+    private final StatementContext ctx;
+
+    public ArrayColumnMapper(Class<?> componentType, StatementContext ctx) {
+        this.componentType = componentType;
+        this.ctx = ctx;
+    }
+
+    @Override
+    public Object[] mapColumn(ResultSet r, int columnNumber, StatementContext ctx) throws SQLException {
+        return buildArray(r.getArray(columnNumber));
+    }
+
+    @Override
+    public Object[] mapColumn(ResultSet r, String columnLabel, StatementContext ctx) throws SQLException {
+        return buildArray(r.getArray(columnLabel));
+    }
+
+    private Object[] buildArray(Array array) throws SQLException {
+        if (UNSUPPORTED_TYPES.contains(array.getBaseType())) {
+            return buildFromResultSet(array);
+        }
+        try {
+            // more efficient, doesn't have to copy an unknown size array
+            return (Object[]) array.getArray();
```

Comment:
It's intended that this only happens for "obvious" mappings that the driver supports, e.g. `int[]` or `String[]` and that it's likely more efficient to let the driver return its array directly rather than materializing it ourselves.  Indeed, this would break if the JDBC driver handles the mapping differently than jdbi does -- maybe this (unproven) performance boost is not worth that possibility.

Or maybe a type whitelist would be more appropriate?